### PR TITLE
Try: make logic for conditional editor_styles consistent

### PR DESCRIPTION
**Pull request by mikachan**
## What?

In https://github.com/WordPress/gutenberg/pull/44640, we added a check for `wp-block-styles` theme support before including the `wp-block-library-theme` CSS in the editor.

This seemed to work well in most cases, however, there are still some cases where this CSS is included. As seen in this issue for TT3 https://github.com/WordPress/twentytwentythree/issues/295, the opinionated width of 100px for the separator is still being loaded from the theme.css file in the editor, but this isn't used on the front end.

| Editor | Front end |
| ----- | ---------- |
| image | image |

## Why?

The editor styles should match the front end.

## How?

This PR is an attempt to never allow the `wp-block-library-theme` CSS to be loaded under the conditions we introduced in https://github.com/WordPress/gutenberg/pull/44640.

I've made the logic consistent between /template-parts-screen.php and /script-loader.php, so the condition to include the theme.css file is only if the active theme supports `wp-block-styles` AND there are no `editor_styles` set.

## Testing Instructions

1. Activate TT3 / a theme that doesn't support `wp-block-styles` AND doesn't include `add_editor_style( 'style.css' );`
2. Insert a Separator block, see the difference in width between the editor and front end

Example separator markup for copy/pasting:

```
<!-- wp:separator -->
<hr class="wp-block-separator has-alpha-channel-opacity"/>
<!-- /wp:separator -->
```

### CODE REVIEW

File: lib/compat/wordpress-6.1/template-parts-screen.php [hunk 1]
```diff
@@ -187,7 +187,7 @@ static function( $classes ) {
 	wp_enqueue_media();
 
 	if (
-		current_theme_supports( 'wp-block-styles' ) ||
+		current_theme_supports( 'wp-block-styles' ) &&
```
Comment (mikachan) [inline lib/compat/wordpress-6.1/template-parts-screen.php:181]: This isn't directly related to the issue this PR is trying to address, but I noticed this logic wasn't quite the same as what is used in `client-assets.php`.

**Comment by scruffian** [comment]
I'm not sure if something changed in TT3 or in Gutenberg, but this isn't working for me, sorry :(

**Comment by mikachan** [comment]
> I'm not sure if something changed in TT3 or in Gutenberg, but this isn't working for me, sorry :(

Thanks for testing this out. I checked this just after you left this comment, and I was seeing the same as you where the fix here didn't have any effect. I tried debugging different things but I didn't get anywhere, so I left it. Now however it does seem to be working again. I'm not sure what's different! But there were several hundred commits between this and trunk, so I've rebased 😅

I think I'm understanding the problem more now, as this seems very specific to TT3. TT3 doesn't have a functions.php, which means it doesn't provide any editor styles. I believe this means this logic, `( ! is_array( $editor_styles ) || count( $editor_styles ) === 0 )`, is causing the theme.css to load in the editor.

~Is this still the logic that we want to determine including the theme.css? I've updated this PR to remove it, which fixes the issue I'm seeing with the separator and means theme.css isn't included.~

~If we still need this check, I think the order needs to be switched so $editor_styles are checked before the theme supports, like this: `( ! is_array( $editor_styles ) || count( $editor_styles ) === 0 ) && current_theme_supports( 'wp-block-styles' )`.~

Nevermind, I don't think this is working either. It's certainly related to the missing `add_editor_style( 'style.css' );` in a functions.php file for TT3, but the above changes I described don't seem to make a difference. If I remove all the calls to the theme.css file in GB, wp-admin/load-styles.php still seems to load these styles in regardless.

**Comment by mikachan** [comment]
I've narrowed this down to the logic for including the theme.css stylesheet not being consistent, namely in script-loader.php. I've updated the logic in this PR to what I think is correct.

Also, I can confirm how to replicate this now. The theme should not support `wp-block-styles` AND not include `add_editor_style( 'style.css' );` / a functions.php file.

However, the only way I could successfully test this was to update the script-loader.php file in Core. I'd appreciate if anyone could confirm that this should be an update to Core rather than in Gutenberg.

I think this update is important to allow themes to not include a functions.php file if it's not necessary (like in Twenty Twenty-Three).

**Comment by carolinan** [comment]
I am not sure what may have changed but I don't see the incorrect editor styles being loaded if I activate Gutenberg **trunk**. ?

**Comment by mikachan** [comment]
Thanks for taking a look, @carolinan. 

I'm still seeing the incorrect width on the separator block on trunk in the editor:

| Editor on trunk | Editor with this PR | Front end |
| ------ | ------ | --------- | 
| image | image | image |

I have to apply the changes in this PR directly to Core (my local version of wordpress-develop) to make this work, rather than just Gutenberg. I'm guessing that's because I'm not sure how the compat files work!

**Comment by carolinan** [comment]
Hm.. I was adding a new separator in the block editor, there it looks fine.
But when I look at the one already saved in the template in the site editor, it is too short.

**Comment by carolinan** [comment]
> I'm not sure how the compat files work!

We need someone to do a zoom call with us and explain it :)

File: lib/compat/wordpress-6.1/template-parts-screen.php (hunk 1 above)
Comment (mikachan) [reply lib/compat/wordpress-6.1/template-parts-screen.php:181]: This needs to be updated in Core in [wp-admin/site-editor.php](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-admin/site-editor.php#L148).

File: lib/compat/wordpress-6.2/script-loader.php [hunk 2]
```diff
@@ -80,7 +80,10 @@ function gutenberg_resolve_assets_override() {
 		'wp-edit-blocks',
 	);
 
-	if ( current_theme_supports( 'wp-block-styles' ) ) {
+	if (
+		current_theme_supports( 'wp-block-styles' ) &&
+		( ! is_array( $editor_styles ) || count( $editor_styles ) === 0 )
+	) {
```
Comment (mikachan) [inline lib/compat/wordpress-6.2/script-loader.php:86]: This needs to be updated in Core in [wp-includes/script-loader.php](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/script-loader.php#L1627).

File: lib/compat/wordpress-6.1/template-parts-screen.php [hunk 3]
```diff
@@ -178,7 +178,7 @@ static function( $classes ) {
...
```
Comment (mikachan) [inline lib/compat/wordpress-6.1/template-parts-screen.php:181]: This needs to be updated in Core in [wp-admin/site-editor.php](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-admin/site-editor.php#L148).

**Comment by jffng** [comment]
As a sanity check, this PR is working for me in the front-end, post editor, and template **part** editor!

Unless you want to write some compatibility code to overwrite what assets are enqueued in the site editor, I do think the easiest fix is to patch this in core as you highlighted.

My impression is that the compat files all work a little differently depending on the context, and the context is always changing... I would definitely join a call with an overview though of the major places where these compat files are necessary and how to approach making changes there.

**Comment by mikachan** [comment]
Thanks @jffng! For completeness (and hopefully to make things easier), I've opened a core patch as well: https://github.com/WordPress/wordpress-develop/pull/3916

**Comment by carolinan** [comment]
I ran into extra padding being added to group blocks that has a background, only in the site editor.
Theme, Twenty Twenty-Three:
I can confirm that the core patch solves this problem.
Before:
Screenshot 2023-02-02 at 06 20 48

After, the themes global padding is applied correctly:
Screenshot 2023-02-02 at 06 34 57

**Comment by carolinan** [comment]
I do not know if this needs to be merged both into Gutenberg and core? If so it needs to be refreshed and moved to the  6.2 folder.

I know that if we want it to be in 6.2 we need to add the backport label..

**Comment by mikachan** [comment]
> I do not know if this needs to be merged both into Gutenberg and core? If so it needs to be refreshed and moved to the 6.2 folder.

Me neither! @ndiego, is this something you could help with? To summarise, this PR is changing files in the lib/compat files, but there is also a [PR open for core](https://github.com/WordPress/wordpress-develop/pull/3916/) to make the same changes there.

**Comment by ndiego** [comment]
> I do not know if this needs to be merged both into Gutenberg and core? If so it needs to be refreshed and moved to the 6.2 folder.
> 
> Me neither! @ndiego, is this something you could help with? To summarise, this PR is changing files in the lib/compat files, but there is also a https://github.com/WordPress/wordpress-develop/pull/3916/ to make the same changes there.

cc @Mamaduka @ntsekouras are you able to provide guidance on this? I am a little unsure myself. Thanks! 🙏

**Comment by glendaviesnz** [comment]
@mikachan my understanding is that if this is to go into 6.2 you would need to:

- Add a new file `lib/compat/wordpress-6.2/template-parts-screen.php`
- Copy your modified `gutenberg_template_parts_screen_init` method into this file and rename it to `gutenberg_template_parts_screen_init_6_2`
- Also add an action to this new file `add_action( 'admin_enqueue_scripts', 'gutenberg_template_parts_screen_init_6_2', 5 );`  - not 100% sure what priority to use here, I think 5 will work as no priority is set on the 6.1 version.
-  Add the new file to the 6.2 section of `lib/load.php`
 - Revert the change that you made in `lib/compat/wordpress-6.1/template-parts-screen.php` 
 - Add the `Backport to WP RC/Beta` label
 - Add details to the relevant sections of https://github.com/WordPress/gutenberg/issues/47187

Sing out if it does get through to the backporting stage. I am happy to help out if you are not familiar with the process.

**Comment by Mamaduka** [comment]
In this case, I think it's okay to modify `gutenberg_template_parts_screen_init` directly.

## Why?

* The custom screen is only available when the Gutenberg plugin is active [with WP 6.0](https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.1/template-parts-screen.php#L40).
* We'll drop WP 6.0 support a few weeks after 6.2 is released, and all code living in `compat/6.1` will be removed.

**Comment by mikachan** [comment]
Thanks, both!

> In this case, I think it's okay to modify gutenberg_template_parts_screen_init directly.

Do you mean directly in Core, like in https://github.com/WordPress/wordpress-develop/pull/3916?

**Comment by Mamaduka** [comment]
Sorry for the late reply, @mikachan!

Re `template-parts-screen.php`: The changes are good to merge as they're 👍

**Comment by mikachan** [comment]
Thanks, @Mamaduka!

So is it OK to merge this PR as it is, or do I need to follow the backporting steps outlined above by @glendaviesnz?

**Comment by mikachan** [comment]
Closing this, as @hellofromtonya kindly tested and merged the Core patch at https://github.com/WordPress/wordpress-develop/pull/3916.

Thanks for everyone's help with this one! 🙇‍♂️

**Pull request closed without merging**